File: script/R/04_pca.R
# ==============================================================================
# PCA — ELLONA/IREN
# Input:  data/processed/TRAIN_FEATURES.csv + output/06_rfecv/rfecv_selected_features.txt
# Output: output/04_pca/pca_scores.png / pca_scree.png / pca_loadings.png
# ==============================================================================
# Richiede: install.packages("ggplot2")
# ==============================================================================

library(ggplot2)

setwd(normalizePath("~/Desktop/IREN"))

# ── Palette ────────────────────────────────────────────────────────────────────
CLASS_COLORS  <- c(ARIA="#4878CF", BIOFILTRO="#6ACC65", BIOGAS="#D65F5F",
                   FORSU="#B47CC7", PERCOLATO="#C4AD66")
CLASS_SHAPES  <- c(ARIA=16, BIOFILTRO=15, BIOGAS=17, FORSU=18, PERCOLATO=8)

out_dir <- "output/04_pca"
dir.create(out_dir, showWarnings=FALSE, recursive=TRUE)

# ── Carica feature selezionate ─────────────────────────────────────────────────
lines_raw <- readLines("output/06_rfecv/rfecv_selected_features.txt")
selected  <- lines_raw[!grepl("^#", lines_raw) & nchar(trimws(lines_raw)) > 0]
cat(sprintf("Feature selezionate: %d\n", length(selected)))

# ── Carica dati ───────────────────────────────────────────────────────────────
df   <- read.csv("data/processed/TRAIN_FEATURES.csv", sep=";", stringsAsFactors=FALSE)
y    <- df$Classe2
X    <- df[, selected, drop=FALSE]

# Imputa NA (mediana per colonna)
for (col in names(X)) {
    if (any(is.na(X[[col]])))
        X[[col]][is.na(X[[col]])] <- median(X[[col]], na.rm=TRUE)
}

# Standardizza
X_scaled <- scale(X)

# ── PCA ───────────────────────────────────────────────────────────────────────
pca_res  <- prcomp(X_scaled, center=FALSE, scale.=FALSE)
scores   <- as.data.frame(pca_res$x)
scores$Classe2   <- y
scores$Sample.ID <- df$Sample.ID

exp_var  <- (pca_res$sdev^2 / sum(pca_res$sdev^2)) * 100
cum_var  <- cumsum(exp_var)
n_comp_90 <- which(cum_var >= 90)[1]

cat(sprintf("\nPC1: %.1f%%  |  PC2: %.1f%%  |  PC3: %.1f%%\n",
            exp_var[1], exp_var[2], exp_var[3]))
cat(sprintf("Componenti per 90%% varianza: %d\n", n_comp_90))

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 1: Scree plot
# ─────────────────────────────────────────────────────────────────────────────
scree_df <- data.frame(
    PC      = 1:min(15, length(exp_var)),
    ExpVar  = exp_var[1:min(15, length(exp_var))],
    CumVar  = cum_var[1:min(15, length(exp_var))]
)

p_scree <- ggplot(scree_df, aes(x=PC, y=ExpVar)) +
    geom_bar(stat="identity", fill="#4878CF", alpha=0.85, width=0.7) +
    geom_line(aes(y=CumVar), color="#D65F5F", linewidth=1, linetype="solid") +
    geom_point(aes(y=CumVar), color="#D65F5F", size=2.5) +
    geom_hline(yintercept=90, linetype="dashed", color="gray50", linewidth=0.8) +
    geom_text(aes(label=sprintf("%.1f%%", ExpVar)), vjust=-0.5, size=3) +
    scale_x_continuous(breaks=1:15) +
    scale_y_continuous(limits=c(0, 105), sec.axis=sec_axis(~., name="Varianza cumulativa (%)")) +
    labs(title="Scree Plot — PCA ELLONA/IREN",
         x="Componente principale",
         y="Varianza spiegata (%)") +
    theme_minimal(base_size=11)

ggsave(file.path(out_dir, "pca_scree.png"), p_scree, width=10, height=5, dpi=150)
cat("✓  pca_scree.png\n")

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 2: PC1 vs PC2 con ellissi di confidenza (95%)
# ─────────────────────────────────────────────────────────────────────────────
p_scores <- ggplot(scores, aes(x=PC1, y=PC2, color=Classe2, shape=Classe2)) +
    stat_ellipse(aes(fill=Classe2), type="norm", level=0.95,
                 geom="polygon", alpha=0.10, linewidth=0.8) +
    geom_point(size=3, alpha=0.9) +
    geom_hline(yintercept=0, color="gray70", linewidth=0.4, linetype="dashed") +
    geom_vline(xintercept=0, color="gray70", linewidth=0.4, linetype="dashed") +
    scale_color_manual(values=CLASS_COLORS) +
    scale_fill_manual(values=CLASS_COLORS) +
    scale_shape_manual(values=CLASS_SHAPES) +
    labs(title="PCA — ELLONA/IREN  (ellissi 95%)",
         x=sprintf("PC1 (%.1f%%)", exp_var[1]),
         y=sprintf("PC2 (%.1f%%)", exp_var[2]),
         color="Classe", shape="Classe", fill="Classe") +
    theme_minimal(base_size=12) +
    theme(legend.position="right",
          panel.grid.minor=element_blank())

ggsave(file.path(out_dir, "pca_scores.png"), p_scores, width=9, height=7, dpi=150)
cat("✓  pca_scores.png\n")

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 3: Loadings PC1 vs PC2 — cerchio delle correlazioni
# ─────────────────────────────────────────────────────────────────────────────
loadings_df <- data.frame(
    Feature = selected,
    Load1   = pca_res$rotation[, 1],
    Load2   = pca_res$rotation[, 2],
    Contrib = sqrt(pca_res$rotation[, 1]^2 + pca_res$rotation[, 2]^2)
)
# Tutte le feature selezionate
top_load <- loadings_df[order(-loadings_df$Contrib), ]

circle_pts <- data.frame(
    x = cos(seq(0, 2*pi, length.out=200)),
    y = sin(seq(0, 2*pi, length.out=200))
)

p_load <- ggplot(top_load) +
    geom_path(data=circle_pts, aes(x=x, y=y),
              color="gray60", linetype="dashed", linewidth=0.7) +
    geom_segment(aes(x=0, y=0, xend=Load1, yend=Load2),
                 arrow=arrow(length=unit(0.25,"cm"), type="closed"),
                 color="#4878CF", linewidth=0.9, alpha=0.85) +
    geom_text(aes(x=Load1*1.12, y=Load2*1.12, label=Feature),
              size=3.2, color="#222222", hjust=0.5, vjust=0.5) +
    geom_hline(yintercept=0, color="gray70", linewidth=0.4) +
    geom_vline(xintercept=0, color="gray70", linewidth=0.4) +
    coord_fixed(xlim=c(-1.15, 1.15), ylim=c(-1.15, 1.15)) +
    labs(title=sprintf("Cerchio delle correlazioni — %d feature RFECV", length(selected)),
         x=sprintf("PC1 (%.1f%%)", exp_var[1]),
         y=sprintf("PC2 (%.1f%%)", exp_var[2])) +
    theme_minimal(base_size=11)

ggsave(file.path(out_dir, "pca_loadings.png"), p_load, width=8, height=8, dpi=150)
cat("✓  pca_loadings.png\n")

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 4: PC1 vs PC3 (vista alternativa)
# ─────────────────────────────────────────────────────────────────────────────
p_13 <- ggplot(scores, aes(x=PC1, y=PC3, color=Classe2, shape=Classe2)) +
    stat_ellipse(aes(fill=Classe2), type="norm", level=0.95,
                 geom="polygon", alpha=0.10, linewidth=0.8) +
    geom_point(size=3, alpha=0.9) +
    geom_hline(yintercept=0, color="gray70", linewidth=0.4, linetype="dashed") +
    geom_vline(xintercept=0, color="gray70", linewidth=0.4, linetype="dashed") +
    scale_color_manual(values=CLASS_COLORS) +
    scale_fill_manual(values=CLASS_COLORS) +
    scale_shape_manual(values=CLASS_SHAPES) +
    labs(title="PCA PC1 vs PC3 — ELLONA/IREN",
         x=sprintf("PC1 (%.1f%%)", exp_var[1]),
         y=sprintf("PC3 (%.1f%%)", exp_var[3]),
         color="Classe", shape="Classe", fill="Classe") +
    theme_minimal(base_size=12)

ggsave(file.path(out_dir, "pca_scores_PC1_PC3.png"), p_13, width=9, height=7, dpi=150)
cat("✓  pca_scores_PC1_PC3.png\n")

# ─────────────────────────────────────────────────────────────────────────────
# Salva scores
# ─────────────────────────────────────────────────────────────────────────────
out_df <- cbind(
    data.frame(Classe2=y, Sample.ID=df$Sample.ID),
    as.data.frame(pca_res$x[, 1:5])
)
write.table(out_df, file.path(out_dir, "pca_results.csv"), sep=";", row.names=FALSE, quote=FALSE)
cat("✓  pca_results.csv\n")

# Summary varianza
cat(sprintf("\nVarianza spiegata:\n"))
for (i in 1:min(8, length(exp_var))) {
    cat(sprintf("  PC%d: %5.1f%%  (cumulativa: %5.1f%%)\n", i, exp_var[i], cum_var[i]))
}

File: script/Python/04_pca.py
"""
PCA — ELLONA/IREN
Input:  data/processed/TRAIN_FEATURES.csv + output/06_rfecv/rfecv_selected_features.txt
Output: output/04_pca/pca_scores.png      (PC1 vs PC2, ellissi per classe)
        output/04_pca/pca_scree.png       (varianza spiegata)
        output/04_pca/pca_loadings.png    (contributo feature a PC1/PC2)
        output/04_pca/pca_3d.png          (PC1 vs PC2 vs PC3)
        output/04_pca/pca_results.csv     (scores + classe)
"""

import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
import os
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.patches import Ellipse
import matplotlib.transforms as transforms

# ── Parametri ──────────────────────────────────────────────────────────────────
TRAIN_FILE    = "data/processed/TRAIN_FEATURES.csv"
FEAT_FILE     = "output/06_rfecv/rfecv_selected_features.txt"
OUT_DIR       = "output/04_pca"
RANDOM_SEED   = 42

CLASS_COLORS = {
    "ARIA":      "#4878CF",
    "BIOFILTRO": "#6ACC65",
    "BIOGAS":    "#D65F5F",
    "FORSU":     "#B47CC7",
    "PERCOLATO": "#C4AD66",
}
CLASS_MARKERS = {
    "ARIA": "o", "BIOFILTRO": "s", "BIOGAS": "^", "FORSU": "D", "PERCOLATO": "P"
}
# ──────────────────────────────────────────────────────────────────────────────

os.makedirs(OUT_DIR, exist_ok=True)

# ── Carica feature selezionate ─────────────────────────────────────────────────
with open(FEAT_FILE) as f:
    selected = [l.strip() for l in f if l.strip() and not l.startswith("#")]
print(f"Feature selezionate: {len(selected)}")

# ── Carica dati ───────────────────────────────────────────────────────────────
df     = pd.read_csv(TRAIN_FILE, sep=";")
y      = df["Classe2"].values
X_raw  = df[selected].values

# Imputa NA e standardizza
X = SimpleImputer(strategy="median").fit_transform(X_raw)
X = StandardScaler().fit_transform(X)

# ── PCA ───────────────────────────────────────────────────────────────────────
pca        = PCA(random_state=RANDOM_SEED)
scores     = pca.fit_transform(X)
exp_var    = pca.explained_variance_ratio_ * 100
cum_var    = np.cumsum(exp_var)
loadings   = pca.components_          # shape: (n_components, n_features)
n_comp_90  = np.argmax(cum_var >= 90) + 1

print(f"\nPC1: {exp_var[0]:.1f}%  |  PC2: {exp_var[1]:.1f}%  |  PC3: {exp_var[2]:.1f}%")
print(f"Componenti per 90% varianza: {n_comp_90}")

# ─────────────────────────────────────────────────────────────────────────────
# Helper: ellisse di confidenza (95%) per una classe
# ─────────────────────────────────────────────────────────────────────────────
def confidence_ellipse(x, y, ax, n_std=2.0, **kwargs):
    if len(x) < 3:
        return
    cov  = np.cov(x, y)
    pearson = cov[0, 1] / np.sqrt(cov[0, 0] * cov[1, 1])
    rx   = np.sqrt(1 + pearson)
    ry   = np.sqrt(1 - pearson)
    ell  = Ellipse((0, 0), width=rx * 2, height=ry * 2, **kwargs)
    sx   = np.sqrt(cov[0, 0]) * n_std
    sy   = np.sqrt(cov[1, 1]) * n_std
    t    = transforms.Affine2D().rotate_deg(45).scale(sx, sy).translate(np.mean(x), np.mean(y))
    ell.set_transform(t + ax.transData)
    ax.add_patch(ell)

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 1: Scree plot
# ─────────────────────────────────────────────────────────────────────────────
fig, axes = plt.subplots(1, 2, figsize=(12, 4))

axes[0].bar(range(1, 11), exp_var[:10], color="#4878CF", alpha=0.8, edgecolor="white")
axes[0].set_xlabel("Componente principale")
axes[0].set_ylabel("Varianza spiegata (%)")
axes[0].set_title("Scree Plot")
axes[0].set_xticks(range(1, 11))
for i, v in enumerate(exp_var[:10]):
    axes[0].text(i+1, v+0.3, f"{v:.1f}%", ha="center", fontsize=8)

axes[1].plot(range(1, len(exp_var)+1), cum_var, "o-", color="#D65F5F", linewidth=1.5, markersize=4)
axes[1].axhline(90, color="gray", linestyle="--", linewidth=1, label="90%")
axes[1].axvline(n_comp_90, color="gray", linestyle="--", linewidth=1)
axes[1].set_xlabel("Numero di componenti")
axes[1].set_ylabel("Varianza cumulativa (%)")
axes[1].set_title("Varianza Cumulativa")
axes[1].legend()
axes[1].set_xlim(1, min(20, len(exp_var)))
axes[1].grid(alpha=0.3)

plt.tight_layout()
plt.savefig(os.path.join(OUT_DIR, "pca_scree.png"), dpi=150)
plt.close()

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 2: PC1 vs PC2 con ellissi
# ─────────────────────────────────────────────────────────────────────────────
fig, ax = plt.subplots(figsize=(9, 7))

classes = sorted(CLASS_COLORS.keys())
for cls in classes:
    mask = y == cls
    if mask.sum() == 0:
        continue
    ax.scatter(scores[mask, 0], scores[mask, 1],
               c=CLASS_COLORS[cls], marker=CLASS_MARKERS[cls],
               s=70, alpha=0.85, edgecolors="white", linewidths=0.5,
               label=f"{cls} (n={mask.sum()})")
    confidence_ellipse(scores[mask, 0], scores[mask, 1], ax,
                       n_std=2.0,
                       edgecolor=CLASS_COLORS[cls],
                       facecolor=CLASS_COLORS[cls],
                       alpha=0.12, linewidth=1.5)

ax.axhline(0, color="gray", linewidth=0.5, linestyle="--")
ax.axvline(0, color="gray", linewidth=0.5, linestyle="--")
ax.set_xlabel(f"PC1 ({exp_var[0]:.1f}%)", fontsize=12)
ax.set_ylabel(f"PC2 ({exp_var[1]:.1f}%)", fontsize=12)
ax.set_title("PCA — ELLONA/IREN  (ellissi 95%)", fontsize=13)
ax.legend(loc="best", framealpha=0.9, fontsize=9)
ax.grid(alpha=0.2)
plt.tight_layout()
plt.savefig(os.path.join(OUT_DIR, "pca_scores.png"), dpi=150)
plt.close()

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 3: Loadings PC1 vs PC2 (biplot delle variabili)
# ─────────────────────────────────────────────────────────────────────────────
load1 = loadings[0]
load2 = loadings[1]

# Tutte le 11 feature selezionate
contrib = np.sqrt(load1**2 + load2**2)
top_idx = np.argsort(contrib)[::-1]

fig, ax = plt.subplots(figsize=(9, 7))
for i in top_idx:
    ax.annotate("", xy=(load1[i], load2[i]), xytext=(0, 0),
                arrowprops=dict(arrowstyle="->", color="#4878CF", lw=1.5))
    ax.text(load1[i] * 1.08, load2[i] * 1.08, selected[i],
            fontsize=8, ha="center", color="#222222")

circle = plt.Circle((0, 0), 1, fill=False, color="gray", linestyle="--", linewidth=0.8)
ax.add_patch(circle)
ax.set_xlim(-1.2, 1.2)
ax.set_ylim(-1.2, 1.2)
ax.axhline(0, color="gray", linewidth=0.5)
ax.axvline(0, color="gray", linewidth=0.5)
ax.set_xlabel(f"PC1 ({exp_var[0]:.1f}%)", fontsize=12)
ax.set_ylabel(f"PC2 ({exp_var[1]:.1f}%)", fontsize=12)
ax.set_title(f"Loadings PCA — {len(selected)} feature selezionate (RFECV)", fontsize=13)
ax.set_aspect("equal")
ax.grid(alpha=0.2)
plt.tight_layout()
plt.savefig(os.path.join(OUT_DIR, "pca_loadings.png"), dpi=150)
plt.close()

# ─────────────────────────────────────────────────────────────────────────────
# PLOT 4: PC1 vs PC2 vs PC3 (3D)
# ─────────────────────────────────────────────────────────────────────────────
from mpl_toolkits.mplot3d import Axes3D  # noqa: F401

fig = plt.figure(figsize=(10, 7))
ax3 = fig.add_subplot(111, projection="3d")

for cls in classes:
    mask = y == cls
    if mask.sum() == 0:
        continue
    ax3.scatter(scores[mask, 0], scores[mask, 1], scores[mask, 2],
                c=CLASS_COLORS[cls], marker=CLASS_MARKERS[cls],
                s=50, alpha=0.85, edgecolors="white", linewidths=0.3,
                label=f"{cls}")

ax3.set_xlabel(f"PC1 ({exp_var[0]:.1f}%)", fontsize=9)
ax3.set_ylabel(f"PC2 ({exp_var[1]:.1f}%)", fontsize=9)
ax3.set_zlabel(f"PC3 ({exp_var[2]:.1f}%)", fontsize=9)
ax3.set_title("PCA 3D — ELLONA/IREN", fontsize=12)
ax3.legend(loc="upper left", fontsize=8, framealpha=0.8)
plt.tight_layout()
plt.savefig(os.path.join(OUT_DIR, "pca_3d.png"), dpi=150)
plt.close()

# ─────────────────────────────────────────────────────────────────────────────
# Salva scores
# ─────────────────────────────────────────────────────────────────────────────
scores_df = pd.DataFrame(scores[:, :5],
                         columns=[f"PC{i+1}" for i in range(5)])
scores_df.insert(0, "Classe2",    y)
scores_df.insert(1, "Sample.ID",  df["Sample.ID"].values)
scores_df.to_csv(os.path.join(OUT_DIR, "pca_results.csv"), sep=";", index=False)

print(f"\n✓  {OUT_DIR}/pca_scree.png")
print(f"✓  {OUT_DIR}/pca_scores.png")
print(f"✓  {OUT_DIR}/pca_loadings.png")
print(f"✓  {OUT_DIR}/pca_3d.png")
print(f"✓  {OUT_DIR}/pca_results.csv")
